### DISCUSSION TITLE
Conda environment best practices

### DISCUSSION THREAD

**Discussion by s-canchi** [nih-cfde/training-and-engagement#366] [Conda Q&A] [answered]
Suppose I have a pipeline with several steps. The steps require different software but they are not incompatible. Is it considered better practice to create one environment for the whole pipeline for simplicity, or a separate environment for each step (e.g. to allow for easy reuse of the individual steps in other pipelines). If the same software is in more than one environment, does it get installed multiple times, or is there a central installation that is reused by different environments?

**Comment by s-canchi** [comment, marked as answer]
- This is a bit of user choice, but in our workflows we use multiple conda environments. It’s actually better because the more software packages you install in a given environment, the slower conda gets when resolving conflicts or updates. 
- The software would all be separate installations. Conda stores installations for each environment in a separate folder.

**Comment by ctb** [comment]
On Thu, Mar 25, 2021 at 02:11:31PM -0700, Saranya Canchi wrote:
> Suppose I have a pipeline with several steps. The steps require different software but they are not incompatible. Is it considered better practice to create one environment for the whole pipeline for simplicity, or a separate environment for each step (e.g. to allow for easy reuse of the individual steps in other pipelines). If the same software is in more than one environment, does it get installed multiple times, or is there a central installation that is reused by different environments?

better practice is to keep environments as small as possible, I think?
but whatever works.

software gets downloaded once, unpacked once, and then (I think) linked in
to different environments. so doesn't take up more space each time it's
installed, etc.

**Reply by marisalim** [reply]
If multiple versions of a software is used - i.e., env1 has software_v1 and env2 has software_v2 - would there be multiple installations of that software?

**Reply by mr-eyes** [reply]
Say you have env1 with python=2.7 and env2 with python=3.6, there will be multiple installations of Python, each in a separate env.

`software` here is associated with the `software version` as well. So, `software = software+version`.
So, if you installed a conda package like `python=3.6` in `env1` and then attempt to install `python=3.6` in env2, conda will search for `python=3.6` in the other envs and **hardlink** it to avoid consuming extra space.

Try executing the following commands for an illustration

```bash
conda create -n env1 python=3.7 # Will download python
conda create -n env2 python=3.7 # Will use the env1 python
```

> _consider using [mamba](https://github.com/mamba-org/mamba)_
